### ISSUE TITLE
Allow names for loss functions and metrics

### ISSUE THREAD

**Issue by mario98** [keras-team/keras#8719]
I have the following scenario: I have a wrapper function for my losses like this:

```
def loss_transformed(target_func, params):
  def func(y_true, y_pred):
    return target_func( transform(y_true, params), y_pred )
  return func
```

Then in model compilation I add the losses like this:

```
losses = {}
for n in N:
  losses['output-%d' % n] = loss_transformed(crossentropy, params[n])
```
_Use case: For example intermediate supervision where _transform_ is a scale function._

However, my loss functions are now named either just `func` if N=1 or `output-n_func` if N>1.

Problem 1: This is a mess. I would like to give intuitive names to the loss functions.

Problem 2: Different loss names for different sizes of N will cause key errors if the loss functions are used for example in checkpoints. Example:

The following only works if N=1 and will crash for N>1:
`checkpoint = ModelCheckpoint(checkpointPath, monitor='val_func')`

while the following only works for N>1 and will crash for N=1:
`checkpoint = ModelCheckpoint(checkpointPath, monitor='val_output-0_func')`

Solution:
Please add some functionality so that we are able to name individual loss functions.
Same applies for metrics.

**Comment by rdiazgar** [comment]
This might do the trick:

https://stackoverflow.com/questions/6527633/how-can-i-make-a-deepcopy-of-a-function-in-python/30714299#30714299?newreg=5b27d70fa16a4e9085e8af24c6acfdae

**Comment by bhavinc** [comment]
Is this issue solved in recent updates?
